TorchSharp/TypeGeneration.tt

```diff
@@ -1307,7 +1626,462 @@ foreach (var type in new [] {
+        
+        /// <summary>
+        ///   Sorts the elements of the tensor along the given dimension. 
+        /// </summary>
+        public System.Tuple<<#=tname#>Tensor,LongTensor> Sort (int dimension, int descending)
+        {
+            var values = new <#=tname#>Tensor ();
+            var indices = new LongTensor ();
+            TH<#=tname#>Tensor_sort (values.handle, indices.handle, this.handle, dimension, descending);
+            return new System.Tuple<<#=tname#>Tensor,LongTensor>(values, indices);
+        }
+
+        [DllImport ("caffe2")]
+        extern static void TH<#=tname#>Tensor_topk (HType values, LongTensor.HType indices, HType self, long k, int dim, int dir, int sorted);
+        
+        /// <summary>
+        ///   Finds the top k of all the elements of the tensor along the given dimension. 
+        /// </summary>
+        public System.Tuple<<#=tname#>Tensor,LongTensor> TopK (long k, int dim, int dir, int sorted)
+        {
+            var values = new <#=tname#>Tensor ();
+            var indices = new LongTensor ();
+            TH<#=tname#>Tensor_topk (values.handle, indices.handle, this.handle, k, dim, dir, sorted);
+            return new System.Tuple<<#=tname#>Tensor,LongTensor>(values, indices);
+        }
+
+        [DllImport ("caffe2")]
+        extern static void TH<#=tname#>Tensor_tril (HType result, HType self, long k);
+        
+        /// <summary>
+        ///   Lower triangle. 
+        /// </summary>
+        public <#=tname#>Tensor TriL (long k)
+        {
+            var result = new <#=tname#>Tensor ();
+            TH<#=tname#>Tensor_tril (result.handle, this.handle, k);
+            return result;
+        }
+
+        [DllImport ("caffe2")]
+        extern static void TH<#=tname#>Tensor_triu (HType result, HType self, long k);
+        
+        /// <summary>
+        ///   Upper triangle. 
+        /// </summary>
+        public <#=tname#>Tensor TriU (long k)
+        {
+            var result = new <#=tname#>Tensor ();
+            TH<#=tname#>Tensor_triu (result.handle, this.handle, k);
+            return result;
+        }
+
+        [DllImport ("caffe2")]
+        extern static void TH<#=tname#>Tensor_cat (HType result, HType ta, HType tb, int dimension);
+        
+        /// <summary>
+        ///   Concatenate tensors along the given dimesion.
+        /// </summary>
+        public <#=tname#>Tensor Concatenate (<#=tname#>Tensor other, int dimension)
+        {
+            var result = new <#=tname#>Tensor ();
+            TH<#=tname#>Tensor_cat (result.handle, this.handle, other.handle, dimension);
+            return result;
+        }
+
+        [DllImport ("caffe2")]
+        extern static void TH<#=tname#>Tensor_catArray (HType result, HType[] ta, int count, int dimension);
+#if false        
+// NOTE: We need to determine the right marshalling for an array of handles.
```

Comment:
I think this is ok, no? What else do you suggest?

Reply:
I tried it and it blew up with a CLR exception, so it was caught before it made it into C++.